cli/flox/src/utils/message.rs

```diff
@@ -112,6 +117,146 @@ pub(crate) fn packages_already_installed(pkgs: &[PackageToInstall], environment_
     }
 }
 
+/// Format a list of overridden fields for an environment.
+fn format_overridden_fields(fields: &[String]) -> String {
+    fields
+        .iter()
+        .map(|key| format!("  - {}", key))
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
+/// Print notices for any environments that have overridden fields during composition.
+pub(crate) fn print_overridden_manifest_fields(lockfile: &Lockfile) {
+    let Some(ref compose) = lockfile.compose else {
+        return;
+    };
+
+    type Field = String;
+    type Environment = String;
+
+    // De-duplicate fields by the last "winning" environment.
+    let winning_env_by_field: BTreeMap<Field, Environment> = compose
+        .warnings
+        .iter()
+        .filter_map(|warning_context| match &warning_context.warning {
+            Warning::Overriding(field) => Some((
+                field.to_string(),
+                warning_context.higher_priority_name.clone(),
+            )),
+            _ => None,
+        })
+        .collect();
+
+    // Invert the de-duplicated map.
+    let mut fields_by_env: BTreeMap<Environment, Vec<Field>> = BTreeMap::new();
+    for (field, env) in winning_env_by_field {
+        fields_by_env.entry(env).or_default().push(field);
+    }
+
+    // Sort the notices by the order that the environments were included and
+    // then the current composer environment (if present) last.
+    let mut messages_by_env: Vec<String> = Vec::new();
+    let ordered_envs = compose.include.iter().map(|include| include.name.clone());
+    for env in ordered_envs {
+        if let Some(fields) = fields_by_env.get(&env) {
+            messages_by_env.push(format!(
+                "- Environment '{}' set:\n{}",
+                env,
+                format_overridden_fields(fields),
+            ));
+        }
+    }
+    if let Some(fields) = fields_by_env.get(COMPOSER_MANIFEST_ID) {
+        messages_by_env.push(format!(
+            "- This environment set:\n{}",
+            format_overridden_fields(fields),
+        ));
+    }
+    if !messages_by_env.is_empty() {
+        let message = formatdoc! {"
+                The following manifest fields were overridden during merging:
+                {}", messages_by_env.join("\n")
+        };
+        info(message);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use flox_rust_sdk::flox::test_helpers::flox_instance;
+    use flox_rust_sdk::models::environment::path_environment::test_helpers::new_path_environment;
+    use flox_rust_sdk::models::environment::Environment;
+    use flox_rust_sdk::utils::logging::test_helpers::test_subscriber_message_only;
+    use indoc::indoc;
+    use pretty_assertions::assert_eq;
+    use tracing::instrument::WithSubscriber;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_print_overridden_manifest_fields() {
```

Comment:
> Changes the unit test to use real environments because it depends more heavily on the implementation of the Compose object; specifically the naming of the environments in Compose.include vs Compose.warnings.

Testing on environments is feeling to me like it's pretty effective. This test takes .01 sec. It's relatively ergonomic to setup. And I think it's not so zoomed in on a single piece of functionality that it misses integration bugs

Reply:
Agree. My only hesitation is that it's just enough test boilerplate that it makes it harder to see the important inputs/outputs and less likely to repeat in smaller variants. But overall I think it's better.